plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/standard/StandardDeployJob.java

```diff
@@ -3,47 +3,45 @@
 import org.eclipse.core.resources.IProject;
 import org.eclipse.core.resources.WorkspaceJob;
 import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Status;
 
 import com.google.cloud.tools.eclipse.appengine.deploy.Messages;
+import com.google.common.base.Preconditions;
 
 public class StandardDeployJob extends WorkspaceJob {
-  
-  private ProjectToStagingExporter projectToStagingExporter;
-  private StagingDirectoryProvider stagingDirectoryProvider;
-  private IProject project;
+
+  private final ProjectToStagingExporter projectToStagingExporter;
+  private final IProject project;
+  private final IPath stagingDir;
 
   public StandardDeployJob(ProjectToStagingExporter exporter,
-                           StagingDirectoryProvider stagingDirectoryProvider,
+                           IPath stagingDir,
                            IProject project) {
     super(Messages.getString("deploy.standard.runnable.name")); //$NON-NLS-1$
+
+    Preconditions.checkNotNull(exporter, "exporter is null");
+    Preconditions.checkNotNull(stagingDir, "stagingDir is null");
+    Preconditions.checkNotNull(project, "project is null");
+
     setRule(project);
+    this.stagingDir = stagingDir;
     projectToStagingExporter = exporter;
-    this.stagingDirectoryProvider = stagingDirectoryProvider;
     this.project = project;
   }
 
   @Override
   public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
-    String stageDir = stagingDirectoryProvider.get();
-    if (stageDir != null) {
-      writeProjectToStageDir(project, stageDir);
-      verifyProjectInStageDir(stageDir);
+    if (stagingDir.toFile().exists()) {
+      projectToStagingExporter.writeProjectToStageDir(project, stagingDir);
       //TODO run stage and deploy operations
       return Status.OK_STATUS;
     } else {
-      return Status.CANCEL_STATUS;
+      return new Status(IStatus.ERROR,
```

Comment:
Yeah, cancel was appropriate until we tried to get the staging directory here from the provider. Since removing that and using a fix temp dir, if that does not exits, it's an error not a cancel event.
The `throws CoreException` is part of the API and it also contains a `Status` object which used by the caller method to report the error.